### PR TITLE
new(ci): use `zig` compiler instead of relying on centos7.

### LINKED ISSUE #3270: CI pipeline uses CentOS 7 which is EOL
The following CI pipeline uses CentOS 7 which is **EOL** now:

https://github.com/falcosecurity/falco/blob/c25ded8f393ddd89f46f330f6e53c806d5bf2423/.github/workflows/reusable_build_packages.yaml#L55

In addition to being EOL, CentOS 7 uses RPM `4.11.x` which uses an old signing method in the RPM binaries it produces which can cause problems when installing Falco in environments with strict security requirements.

Relevant info in the RPM release notes: https://rpm.org/wiki/Releases/4.14.0

It would be great if the pipeline could be updated to some current RPM-based distro (no strong opinions on the specific distro as long as it contains RPM >= `4.14.x` which uses the modern signing method).

Thanks! :pray:

### PR DESCRIPTION
**Pull request by FedeDP**
**What type of PR is this?**

/kind cleanup

**Any specific area of the project related to this PR?**

/area CI

**What this PR does / why we need it**:

This PR drops centos7 from our CI by instead relying on zig compiler to provide glibc 2.17 compatible builds for us.

**Which issue(s) this PR fixes**:

Fixes #3270

**Special notes for your reviewer**:

Linked libs PRs:
* https://github.com/falcosecurity/libs/pull/2034 -> new c-ares builds with zig 
* https://github.com/falcosecurity/libs/pull/2036 -> libanl is not present under zig; fixes build
* https://github.com/falcosecurity/libs/pull/2037 -> title says it all
* https://github.com/falcosecurity/libs/pull/2043 -> avoids Falco linking issue `undefined symbol: google::protobuf::internal::InternalMetadata::~InternalMetadata()`
* https://github.com/falcosecurity/libs/pull/2045 -> fixes arm64 `Trace/breakpoint trap` issue for threadinfo and k8saudit plugin opening (source_plugin.c UB)
* https://github.com/falcosecurity/libs/pull/2049 -> cmake patches for libelf and grpc to support zig build
* https://github.com/falcosecurity/libs/pull/2088 -> more UBs fixes for libs

TODO:
- [x] sanitizer build fails at cmake configure step, because of: https://github.com/ziglang/zig/issues/11403. **Workaround: disable `zig` for sanitizer builds, ie: only use it for to-be-shipped artifacts.**
- [x] arm64 build is failing while building abseil randen_hwaes.cc file, since their cmake tries to detect platform features and enforces `-march=armv8-a+crypto`, but `armv8` is unsupported by zig. **Workaround: patch https://github.com/abseil/abseil-cpp/blob/1031858eb2b87a5d47f236d8eac11d8b05c535d5/absl/copts/GENERATED_AbseilCopts.cmake#L225 to use a supported arch name.** 
- [x] fix `test-dev-packages-arm64` -> it seems like a `Trace/breakpoint trap` is caught during `sinsp::open...` methods calls:
```
Tue Sep  3 10:46:35 2024: Loaded event sources: syscall
Tue Sep  3 10:46:35 2024: Enabled event sources: syscall
Tue Sep  3 10:46:35 2024: Opening 'syscall' source with Kernel module
Trace/breakpoint trap
```
- [x] fix `test-dev-packages-arm64` -> it seems like a `Trace/breakpoint trap` is caught related to stats_writer line: `output_fields["falco.host_boot_ts"] = machine_info->boot_ts_epoch;` -> bca6f31e94f5417ecf07975ab65e520c46656fac
- [x] fix `test-dev-packages-arm64` -> it seems like a `Trace/breakpoint trap` is caught related to k8saudit opening 
- [x] libelf is built bundled

**Does this PR introduce a user-facing change?**:

```release-note
new(ci): use `zig` compiler instead of relying on centos7.
```

### CODE REVIEW

**Comment by FedeDP** [comment]
So, we need to disable the build with shared libelf, because `zig` statically links all specified libraries; see: https://github.com/ziglang/zig/issues/7094
Indeed the CI complains it cannot find system libelf.
Since this was an :hot_pepper: topic for the CNCF, i am going to hold this PR; at the same time, i think we should actually find a way to drop centos:7 while maintaining same glibc requirements.


Anyway, even trying with BUNDLED_LIBELF, gives a build error...

/hold

File: submodules/falcosecurity-rules
Comment (FedeDP) [inline submodules/falcosecurity-rules:1]: As always, a `git commit -sam` committed this :rofl:

File: .github/workflows/reusable_build_packages.yaml
Comment (FedeDP) [inline .github/workflows/reusable_build_packages.yaml:90]: Enabled bundled libelf to build with zig. Of course this is not desiderable.

**Comment by FedeDP** [comment]
Now failing the build on 
> ld.lld: error: undefined symbol: arc4random_buf
>>> referenced by ares_rand.c:270

That's probably because c-ares does some magic to detect that symbol: https://github.com/c-ares/c-ares/blob/main/CMakeLists.txt#L456; that symbol was added in glibc2.36; most probably it is detecting the symbol even if `zig` is not actually supporting it since we are targeting a build with glibc-2.17.

**Comment by FedeDP** [comment]
I tried to build `c-ares` from scratch (ie: without Falco involved at all) with `zig` and it worked fine with the same `export`s we are doing in the CI.

**Comment by FedeDP** [comment]
Update: i saw that updating c-ares to 1.33.1 (latest release) fixed the issue. I am going to bump it in libs (if possible, ie: if grpc does not complain :/ )

**Comment by FedeDP** [comment]
So, i now bumped this branch to use libs [`chore/zig_build` branch HEAD](https://github.com/falcosecurity/libs/tree/chore/zig_build); that is a branch i pushed upstream with all zig related fixes on it:
* https://github.com/falcosecurity/libs/pull/2034
* https://github.com/falcosecurity/libs/pull/2036
* https://github.com/falcosecurity/libs/pull/2037

Moreover, i found out that CMAKE_{C,CXX}_COMPILER truncates `CC` and `CXX` strings at first whitespace, leading with cmake using `zig` as compiler for both, instead of `zig cc` and `zig c++`.
I created 2 small wrapper scripts in CI to workaround this issue:
```
cat > zig-linux-${{ inputs.arch }}-${ZIG_VERSION}/zig-cc << EOF
#!/bin/bash
zig cc -target ${{ inputs.arch }}-linux-gnu.2.17 "$@"
EOF
chmod +x zig-linux-${{ inputs.arch }}-${ZIG_VERSION}/zig-cc
echo "CC=zig-cc" >> $GITHUB_ENV

// same for zig-c++
```

**Comment by FedeDP** [comment]
So, we now have artifacts working (passing tests too!); here are the differences for arm64:
* normal Falco:
```
ldd falco
linux-vdso.so.1 (0x0000e4ff9035f000)
	libelf.so.1 => /lib/aarch64-linux-gnu/libelf.so.1 (0x0000e4ff902f0000)
	librt.so.1 => /lib/aarch64-linux-gnu/librt.so.1 (0x0000e4ff902c0000)
	libanl.so.1 => /lib/aarch64-linux-gnu/libanl.so.1 (0x0000e4ff90290000)
	libdl.so.2 => /lib/aarch64-linux-gnu/libdl.so.2 (0x0000e4ff90260000)
	libpthread.so.0 => /lib/aarch64-linux-gnu/libpthread.so.0 (0x0000e4ff90230000)
	libz.so.1 => /lib/aarch64-linux-gnu/libz.so.1 (0x0000e4ff901f0000)
	libstdc++.so.6 => /lib/aarch64-linux-gnu/libstdc++.so.6 (0x0000e4ff8fe00000)
	libm.so.6 => /lib/aarch64-linux-gnu/libm.so.6 (0x0000e4ff90140000)
	libgcc_s.so.1 => /lib/aarch64-linux-gnu/libgcc_s.so.1 (0x0000e4ff90100000)
	libc.so.6 => /lib/aarch64-linux-gnu/libc.so.6 (0x0000e4ff8fc40000)
	/lib/ld-linux-aarch64.so.1 (0x0000e4ff90322000)
	libzstd.so.1 => /lib/aarch64-linux-gnu/libzstd.so.1 (0x0000e4ff8fb80000)

objdump -T  falco | grep -Eo 'GLIBC_\S+' | sort -u
GLIBC_2.17)

du -h falco
26M	falco
```

* zig Falco:
```
ldd falco
linux-vdso.so.1 (0x0000e17bf7aff000)
	libm.so.6 => /lib/aarch64-linux-gnu/libm.so.6 (0x0000e17bf7a20000)
	libpthread.so.0 => /lib/aarch64-linux-gnu/libpthread.so.0 (0x0000e17bf79f0000)
	libc.so.6 => /lib/aarch64-linux-gnu/libc.so.6 (0x0000e17bf7830000)
	libdl.so.2 => /lib/aarch64-linux-gnu/libdl.so.2 (0x0000e17bf7800000)
	librt.so.1 => /lib/aarch64-linux-gnu/librt.so.1 (0x0000e17bf77d0000)
	/lib/ld-linux-aarch64.so.1 (0x0000e17bf7ac2000)
	
objdump -T falco | grep -Eo 'GLIBC_\S+' | sort -u
GLIBC_2.17)

du -h falco
124M	falco
```
So, the zig Falco:
* links less things, and that's expected since libelf is now statically built, libanl is no more linked, there is no libgcc/libstdc++ and libzstd; moreover libz dynamically linked by libelf is gone too.
* has same glibc requirements (2.17+)
* is much bigger (at least in the Debug build of arm64; the x86_64 debug build is much similar, 29M to 24M).

File: .github/workflows/reusable_build_packages.yaml
Comment (FedeDP) [inline .github/workflows/reusable_build_packages.yaml:72]: Needed to find system libelf.

File: cmake/modules/falcosecurity-libs-repo/patches/zig_libelf.patch
Comment (FedeDP) [inline cmake/modules/falcosecurity-libs-repo/patches/zig_libelf.patch:1]: Basically, we need to patch libs libelf cmake module to:
* add the `HINTS` about the possible location of the libelf library. Normally, cmake would auto-detect that it is running under debian/ubuntu and add that folder to its search list; under zig that is not true and we need to specifically add it
* to avoid including the whole `/usr/include`, we just copy libelf headers to `CMAKE_CURRENT_BINARY_DIR` and include it. This is needed to avoid including the whole `/usr/include` during the compilation that would lead to compilation issues since we would mix the system toolchain with the zig one

File: cmake/modules/falcosecurity-libs-repo/patches/zig_aarch64_grpc.patch
Comment (FedeDP) [inline cmake/modules/falcosecurity-libs-repo/patches/zig_aarch64_grpc.patch:1]: Abseil-cpp is smart enough to detect certain hw capabilities to improve its perf, in this specific case we are talking about the `ABSL_RANDOM_HWAES_ARM64_FLAGS`: https://github.com/abseil/abseil-cpp/blob/master/absl/copts/GENERATED_AbseilCopts.cmake#L225.
Setting `armv8` is not a supported `march` for zig; therefore we patch the grpc cmake module (that builds abseil-cpp) to PATCH aforementioned abseil source file to name that `march` in another, zig-compatible, way.

File: cmake/modules/falcosecurity-libs.cmake
Comment (FedeDP) [inline cmake/modules/falcosecurity-libs.cmake:59]: HAS_CAPTURE was dropped since ages :D

File: userspace/falco/stats_writer.cpp
Comment (FedeDP) [inline userspace/falco/stats_writer.cpp:341]: As weird as it seems, this stupid cast fixed a SIGTRAP generated on debug builds by zig.
If i removed the `pragma pack` from the scap_machine_info struct, the issue went away. Most probably this is just a bug somewhere in how zig/clang compile the sources.

File: cmake/modules/falcosecurity-libs-repo/patches/CMakeLists.txt
Comment (FedeDP) [inline cmake/modules/falcosecurity-libs-repo/patches/CMakeLists.txt:22]: libelf patch uses `@CMAKE_SYSTEM_PROCESSOR@` thus it needs to be configured.
`@ONLY` to avoid touching also normal cmake variables like `${FOO}`.

File: cmake/modules/CompilerFlags.cmake
Comment (FedeDP) [inline cmake/modules/CompilerFlags.cmake:1]: Can revert this one but it is not a nice behavior to erase pre-defined `CMAKE_{C,CXX}_FLAGS` IMHO.

**Comment by FedeDP** [comment]
The PR is ready; idea is to merge this right after the Falco release, after https://github.com/falcosecurity/libs/pull/2034 and https://github.com/falcosecurity/libs/pull/2037 are merged in Falco libs (i'll then take care of update this PR to use libs master of course).

Then, we will have 4 months to discover any new issue and fix it.
Also, as soon as i'll have some spare time, i'd like to benchmark the new binary (eg: by locally building sinsp with zig and running `make bench` on it, since we now have a bunch of google benchmarks in libs, and doing the same from Falco master).

**Comment by FedeDP** [comment]
/unhold
Everything is fixed and we are using libs master.

**Comment by FedeDP** [comment]
/hold
Don't get why 
>  CI Build / test-dev-packages-arm64 / test-packages (pull_request) Failing after 4m 

is failing on 2 tests because cannot find `.scap` files :/

**Review by leogr — APPROVED** [review]

**Review by leogr — APPROVED** [review]

**Comment by leogr** [comment]
> This PR may bring feature or behavior changes in the Falco engine and may require the engine version to be bumped.
> 
> Please double check **userspace/engine/falco_engine_version.h** file. See [versioning for FALCO_ENGINE_VERSION](https://github.com/falcosecurity/falco/blob/master/RELEASE.md#falco-repo-this-repo).
> 
> /hold

False positive

**Comment by FedeDP** [comment]
Yep, tomorrow will rebase on latest libs master that fixes arm64 test-dev-packages and unhold this one :)

**Comment by FedeDP** [comment]
/unhold
Ready!

**Review by leogr — APPROVED** [review]

**Comment by FedeDP** [comment]
:loudspeaker: it might happen that master build fails because the cached key (ie: zig version) is no more available and `zig` moved to a newer master version, thus we cannot find the correct zig version to be downloaded.
In that case, i'll update zig to latest in libs and restart the CI here.

**Comment by FedeDP** [comment]
/hold 
until next week.

**Review by jasondellaluce — APPROVED** [review]

**Comment by FedeDP** [comment]
I'll keep this on hold until we release Falco 0.39.1 to avoid issues with the new CI.

**Comment by FedeDP** [comment]
/unhold
Let's see if master CI complains (i guess so :D )

**Pull request merged**
